src/core/analyzer-cache-context.ts

```diff
@@ -0,0 +1,483 @@
+      throw new Error(`Internal error: document ${resolvedUrl} already exists`);
+    }
+
+    const document = new Document(scannedDocument, this);
+    if (!this._cache.analyzedDocumentPromises.has(resolvedUrl)) {
+      this._cache.analyzedDocumentPromises.set(
+          resolvedUrl, Promise.resolve(document));
+    }
+    this._cache.analyzedDocuments.set(resolvedUrl, document);
+    document.resolve();
+    return document;
+  }
+
+  /**
+   * Gets an analyzed Document from the document cache. This is only useful for
+   * Analyzer plugins. You almost certainly want to use `analyze()` instead.
+   *
+   * If a document has been analyzed, it returns the analyzed Document. If not
+   * the scanned document cache is used and a new analyzed Document is returned.
+   * If a file is in neither cache, it returns `undefined`.
+   */
+  _getDocument(url: string): Document|undefined {
+    const resolvedUrl = this._resolveUrl(url);
+    let document = this._cache.analyzedDocuments.get(resolvedUrl);
+    if (document) {
+      return document;
+    }
+    const scannedDocument = this._cache.scannedDocuments.get(resolvedUrl);
+    return scannedDocument && this._makeDocument(scannedDocument);
+  }
+
+  /**
+   * This is only useful for Analyzer plugins.
+   *
+   * If a url has been scanned, returns the ScannedDocument.
+   */
+  _getScannedDocument(url: string): ScannedDocument|undefined {
+    const resolvedUrl = this._resolveUrl(url);
+    return this._cache.scannedDocuments.get(resolvedUrl);
+  }
+
+  async getTelemetryMeasurements(): Promise<Measurement[]> {
+    return this._telemetryTracker.getMeasurements();
+  }
+
+  /**
+   * Clear all cached information from this analyzer instance.
+   *
+   * Note: if at all possible, instead tell the analyzer about the specific
+   * files that changed rather than clearing caches like this. Caching provides
+   * large performance gains.
+   */
+  clearCaches(): AnalyzerCacheContext {
+    return this._fork(new AnalysisCache());
+  }
+
+  /**
+   * Return a copy, but with the given cache.
+   */
+  private _fork(cache: AnalysisCache): AnalyzerCacheContext {
+    const copy = new AnalyzerCacheContext({
+      lazyEdges: this._lazyEdges,
+      parsers: this._parsers,
+      scanners: this._scanners,
+      urlLoader: this._loader,
+      urlResolver: this._resolver
+    });
+    copy._telemetryTracker = this._telemetryTracker;
+    copy._cache = cache;
+    copy._generation = this._generation + 1;
+    return copy;
+  }
+
+  /**
+   * Scan a toplevel document given its url and optionally its contents.
+   */
+  private async _scan(
+      resolvedUrl: string, contents?: string,
+      visited?: Set<string>): Promise<ScannedDocument|'visited'> {
+    if (visited && visited.has(resolvedUrl)) {
+      return 'visited';
+    }
+    visited = visited || new Set();
+    visited.add(resolvedUrl);
+    const cachedResult = this._cache.scannedDocumentPromises.get(resolvedUrl);
```

Comment:
Oh, yeah that's easier to reason about, +1.

And +1 to typescript for correctly determining that `Promise<T> | undefined` awaits to `T | undefined`